From 9e818c2fa4efc42a077f193068312b8a287982b3 Mon Sep 17 00:00:00 2001
From: Vivek Chettri <vchettri@qti.qualcomm.com>
Date: Mon, 21 Jul 2025 16:24:23 +0530
Subject: [PATCH] UPSTREAM:PROTOCOL: hostapd: Configure nftables for EPCS

Configure NFT rules for EPCS enabled clients.
The EPCS client mac based NFT rule will help priortize the
the flows of the client using the underlying QoS framework.

Patch-dependency:
r85-01-UPSTREAM-PROTOCOL-HOSTAPD-Introduce_QoS_Management_cmd_and_attr.patch
r85-02-UPSTREAM-PROTOCOL-HOSTAPD-Define-QoS-Management-definitions.patch
r85-03-UPSTREAM-PROTOCOL-CP-HOSTAPD-QM-NL80211-request-and-response.patch
r79-0011-UPSTREAM-PROTOCOL-hostapd-Send-EPCS-Action-frames.patch

Tag: EPCS

Signed-off-by: Vivek Chettri <vchettri@qti.qualcomm.com>
---
 src/ap/ieee802_11.h     |  3 +++
 src/ap/ieee802_11_eht.c | 29 +++++++++++++++++++++++++++--
 src/ap/sta_info.c       |  3 +++
 3 files changed, 33 insertions(+), 2 deletions(-)

diff --git a/src/ap/ieee802_11.h b/src/ap/ieee802_11.h
index 0980ad3..02870bf 100644
--- a/src/ap/ieee802_11.h
+++ b/src/ap/ieee802_11.h
@@ -401,6 +401,9 @@ int hostapd_process_assoc_ml_info(struct hostapd_data *hapd,
 
 #ifdef CONFIG_IEEE80211BE
 void hostapd_epcs_timeout_handler(void *eloop_ctx, void *timeout_ctx);
+int hostapd_configure_epcs(struct hostapd_data *hapd,
+				  struct sta_info *sta,
+				  enum qos_mgmt_req_type req_type);
 #endif /* CONFIG_IEEE80211BE */
 u8 * hostapd_fragment_multi_link_element(struct wpabuf *buf, u8 *pos);
 #endif /* IEEE802_11_H */
diff --git a/src/ap/ieee802_11_eht.c b/src/ap/ieee802_11_eht.c
index a134648..e250d29 100644
--- a/src/ap/ieee802_11_eht.c
+++ b/src/ap/ieee802_11_eht.c
@@ -21,6 +21,7 @@
 #include "utils/eloop.h"
 #include "wmm.h"
 #include "ap/ctrl_iface_ap.h"
+#include <linux/netfilter.h>
 
 static u16 ieee80211_eht_ppet_size(u16 ppe_thres_hdr, const u8 *phy_cap_info)
 {
@@ -1924,12 +1925,13 @@ void hostapd_epcs_timeout_handler(void *eloop_ctx, void *timeout_ctx)
 	mld_info->epcs.timer_started = false;
 }
 
-static int hostapd_configure_epcs(struct hostapd_data *hapd,
+int hostapd_configure_epcs(struct hostapd_data *hapd,
 				  struct sta_info *sta,
 				  enum qos_mgmt_req_type req_type)
 {
 	struct qm_req_data qm_req = {0};
 	struct qm_resp_data qm_resp = {0};
+	struct hostapd_nft_rule_params rule = {0};
 
 	if (hapd->driver == NULL)
 		return -1;
@@ -1960,12 +1962,24 @@ static int hostapd_configure_epcs(struct hostapd_data *hapd,
 		return -1;
 	}
 
+	/* Program NFT rule delete */
+	if (req_type == QM_REMOVE_REQ) {
+		os_snprintf(rule.chain, sizeof(rule.chain), "%s_%s", CHAIN_NAME,
+			    hapd->conf->iface);
+		os_snprintf(rule.table, sizeof(rule.table), "%s", TABLE_NAME);
+		rule.mark = (EPCS_QM_ID << 8) | HOSTAPD_QOS_SCS_TAG;
+		rule.nf_family = NFPROTO_NETDEV;
+		memcpy(rule.dmac, sta->addr, ETH_ALEN);
+		hostapd_ucode_config_nft_rule(hapd, &rule, false);
+	}
+
 	/* Send the QoS request */
 	if (hapd->driver->set_qos(hapd->drv_priv, &qm_req, &qm_resp)) {
 		wpa_printf(MSG_ERROR, "Set QoS failed");
 		return -1;
 	}
 
+
 	if (qm_resp.qm_resp_desc[0].qm_id != qm_req.qm_req_desc[0].qm_id) {
 		wpa_printf(MSG_ERROR, "EPCS QM ID mismatch in response:%u",
 			   qm_resp.qm_resp_desc[0].qm_id);
@@ -1978,6 +1992,18 @@ static int hostapd_configure_epcs(struct hostapd_data *hapd,
 		return -1;
 	}
 
+	/* Program NFT rule add */
+	if (req_type == QM_ADD_REQ) {
+		os_snprintf(rule.chain, sizeof(rule.chain), "%s_%s", CHAIN_NAME,
+			    hapd->conf->iface);
+		os_snprintf(rule.table, sizeof(rule.table), "%s", TABLE_NAME);
+		memcpy(rule.dmac, sta->addr, ETH_ALEN);
+		rule.valid_flags |= NFT_RULE_PARAM_DMAC;
+		rule.mark = (EPCS_QM_ID << 8) | HOSTAPD_QOS_SCS_TAG;
+		rule.nf_family = NFPROTO_NETDEV;
+		hostapd_ucode_config_nft_rule(hapd, &rule, true);
+	}
+
 	return 0;
 }
 
@@ -2091,7 +2117,6 @@ failed:
 			   "for sta with mld addr" MACSTR, prev_epcs_state, epcs_info->action_code, sta_epcs->state,
 			   MAC2STR(sta->mld_info.common_info.mld_addr));
 		return -1;
-
 	} else {
 		if (sta_epcs->state == EPCS_STATE_DISABLED)
 			ret = hostapd_configure_epcs(hapd, sta, QM_REMOVE_REQ);
diff --git a/src/ap/sta_info.c b/src/ap/sta_info.c
index c3c202a..b28b5bc 100644
--- a/src/ap/sta_info.c
+++ b/src/ap/sta_info.c
@@ -563,6 +563,9 @@ void ap_free_sta(struct hostapd_data *hapd, struct sta_info *sta)
 	os_free(sta->ifname_wds);
 
 #ifdef CONFIG_IEEE80211BE
+
+	wpa_printf(MSG_INFO, "EPCS Tear Down from AP free STA");
+	hostapd_configure_epcs(hapd, sta, QM_REMOVE_REQ);
 	eloop_cancel_timeout(hostapd_epcs_timeout_handler, hapd, sta);
 	ap_sta_free_sta_profile(&sta->mld_info);
 #endif /* CONFIG_IEEE80211BE */
-- 
2.34.1

